### DISCUSSION TITLE
Development diary

### DISCUSSION THREAD

**Discussion by skalarproduktraum** [scenerygraphics/scenery#357] [Show and tell]
Let's use this here as a developer diary!

**Comment by skalarproduktraum** [comment]
### Debugging screen configurations for distributed setups

scenery has now gained a "debugger" for screen configurations for CAVEs, here's what the configuration at the Center for Systems Biology Dresden looks like:
![screenconfig-csbd](https://user-images.githubusercontent.com/586495/101773501-26599f80-3aed-11eb-8b0b-bbc460988d70.gif)

And here's another system we are just setting up, the CAVE at the [LANEvol Lab in Geneva](https://www.lanevol.org):
![screenconfig-geneva](https://user-images.githubusercontent.com/586495/101773511-2e194400-3aed-11eb-8b40-b9051fe458cc.gif)

This helps with ending up with something like this:
![VR in Science](https://user-images.githubusercontent.com/586495/101773625-56a13e00-3aed-11eb-9c56-4caf74174779.jpg) (Thanks for the photo to Katrin Boes, MPI-CBG Dresden)

**Comment by skalarproduktraum** [comment]
### SteamVR/OpenVR improvements

Today's entry is about improvements in handling OpenVR/SteamVR devices. We have significantly improved the support for Oculus devices, such as the Quest 2, which works just fine over Oculus Link with scenery. `OpenVRHMD` now supports all buttons exposed by SteamVR, and joystick/trackpad handling is now consistent between HTC Vive and other devices. In addition, we now also track velocity and angular velocity of all devices. Adding behaviours and triggering them with buttons is now also much nicer than before, and can be achieved by calling `hmd.addKeyBinding("toggle_boxes", TrackerRole.LeftHand, OpenVRHMD.OpenVRButton.Side)` instead of having to dig out the keycode the OpenVR button is bound to. Finally, we have fixed an issue where key bindings were not triggered due to a recent change in @tpietzsch's awesome [ui-behaviour library](https://github.com/scenerygraphics/scenery) we use for input handling and key bindings. Thanks to @ljjh20 for the report 👍

I have also expanded the `VRControllerExample`, there you can now nudge boxes around. If you are curious about key bindings on VR controllers, and testing them against scene objects, check out that example.

![VRControllerExample_Nudging](https://user-images.githubusercontent.com/586495/102378350-fff0a400-3fc5-11eb-818a-c5f92e9adbcf.gif)

**Comment by skalarproduktraum** [comment]
### Rendering proteins from PDBs

Thanks to @Domino2357, scenery has recently gained support for creating protein geometries from PDB files, and can also download them automagically from the internet. Here we overdid it a bit:

![image](https://user-images.githubusercontent.com/586495/102540866-f6dc0180-40af-11eb-90a3-0b12e088eacc.png)

In this image, you can see two instances of the insect wing muscle protein 2w49, two instances of the actin filament 3g37, and one instance of the thermosome 3los. About 44000 drawcalls in total, so this still needs a bit of optimisation. Great work, @Domino2357!

PS: The above can already be used in [sciview](https://github.com/scenerygraphics/sciview)! Check the `Edit > Add Protein from PDB....` command.

**Comment by elect86** [comment]
imgui kind of usable

inputs are glfw-only, and I had to overcome the scenery limitation of not rendering garbage in vertex buffers by clearing them before pushing new small valid data inside, but so far it works

![Screenshot from 2021-03-24 16-20-56](https://user-images.githubusercontent.com/5020626/112336004-fbbadd80-8cbc-11eb-8487-e2d97976f355.png)
